### PR TITLE
Offset

### PR DESCRIPTION
**Pull request by jjimenezshaw**
I need to synchronize, not the centers of both maps, but other point. To create a continuous map, or a mosaic, or for any other purpose.

On this pull request the user can pass a function `offsetFn` that defines this transformation. It is very generic to allow flexibility. However, the most common use is really easy with the offsetHelper factory (as explained in the README.md).

Now, on each setView or similar function that resets a center, this center is computed with the provided function.

The default value for this option is an identity function, that is, returns the same center.

### CODE REVIEW

File: L.Map.Sync.js [hunk 1]
```diff
@@ -41,6 +48,27 @@
             return this;
         },
 
+        // Helper function to compute the offset easily
+        // The arguments are relative positions respect refence and target maps
```
Comment (jieter) [inline L.Map.Sync.js]: Sentence seems to miss some (parts of) words (added with **emphasis**):
  ... are relative positions **with** respect to ref**er**ense and target maps.

File: L.Map.Sync.js [hunk 2]
```diff
@@ -41,6 +48,27 @@
...
+        // The values can be less than 0 or greater than 1. It will sync
+        // points out of the map.
+        offsetHelper: function (ratioRef, ratioTgt) {
+            var or = L.Util.isArray(ratioRef) ? ratioRef : [0.5, 0.5]
+            var ot = L.Util.isArray(ratioTgt) ? ratioTgt : [0.5, 0.5]
+            return function (center, zoom, refMap, tgtMap) {
+                var rs = refMap.getSize();
+                var ts = tgtMap.getSize();
+                var pt = refMap.project(center, zoom)
+                               .subtract([(0.5-or[0])*rs.x, (0.5-or[1])*rs.y])
```
Comment (jieter) [inline L.Map.Sync.js]: Please conform to project style: use spaces around operators

File: README.md [hunk 3]
```diff
@@ -36,6 +36,21 @@ mapB.sync(mapC);
...
+You can synchronize not only the centers, but other points, using the option `offsetFn`.
+The parameters send to the function are `(center, zoom, referenceMap, targetMap)`, and it must return the equivalent center to produce your offset. That means, the center to pass to setView.
+
+In most cases, you can use the factory `offsetHelper`, that accepts two arrays of two elements each `(ratioRef, ratioTgt)`. The meaning of this array is the relative position -relative to the top left corner and the whole size- in the map container of the point to synchronize. (1 is for the whole width or height). Values greater than 1 or less than 0 work fine.
+
+For instance `mapB.sync(mapC, {offsetFn: L.Map.prototype.offsetHelper([0, 1], [1, 1])});` will sync the bottom left corner `[0, 1]` in the reference map (mapB) with the bottom right corner `[1, 1]` in the target map (mapC).
+
+As well `mapB.sync(mapA, {offsetFn: mapB.offsetHelper([0, 0], [1, 0.5])});` will sync the top left corner `[0 ,0]` in mapB with the center of the right side `[1, 0.5]` in mapA.
+
+If you want the actions to be synced vice-versa, you should use simetric values (as reference and target are swapped).
```
Comment (jieter) [inline README.md]: simetric -> symmetric.

I think this line could use an example of a non-center, symmetric sync pair.

File: README.md [hunk 4]
```diff
@@ -36,6 +36,21 @@ mapB.sync(mapC);
...
+
+The default behaviour uses `[0.5, 0.5], [0.5, 0.5]`, that synchronizes the centers.
```
Comment (jieter) [inline README.md]: I would rephrase:

The default behaviour is to synchronize the centers, the corresponding offset points are `[0.5, 0.5], [0.5, 0.5]`.

or something like that.

File: L.Map.Sync.js [hunk 5]
```diff
@@ -41,6 +48,27 @@
...
 
+        // Helper function to compute the offset easily
+        // The arguments are relative positions respect refence and target maps
+        // of the point to sync.
+        // If you provide ratioRef=[0,1], ratioTgt=[1,0] will sync
+        // the bottom left corner of the reference map
+        // with the top right corner of the target map.
+        // The values can be less than 0 or greater than 1. It will sync
+        // points out of the map.
+        offsetHelper: function (ratioRef, ratioTgt) {
```
Comment (jieter) [inline L.Map.Sync.js]: Not sure if this method should be at home in this location, I mean, it doesn't interact with the instance at all. Is `L.Util` a better location?

File: L.Map.Sync.js (hunk 1 above)
Comment (jjimenezshaw) [inline L.Map.Sync.js]: Thanks

File: README.md [hunk 6]
```diff
@@ -36,6 +36,21 @@ mapB.sync(mapC);
 mapC.sync(mapA);
 mapC.sync(mapB);
 ```
+## Offset
+You can synchronize not only the centers, but other points, using the option `offsetFn`.
+The parameters send to the function are `(center, zoom, referenceMap, targetMap)`, and it must return the equivalent center to produce your offset. That means, the center to pass to setView.
+
+In most cases, you can use the factory `offsetHelper`, that accepts two arrays of two elements each `(ratioRef, ratioTgt)`. The meaning of this array is the relative position -relative to the top left corner and the whole size- in the map container of the point to synchronize. (1 is for the whole width or height). Values greater than 1 or less than 0 work fine.
+
```
Comment (jieter) [inline README.md]: From this paragraph, I wonder:
 - can you add/is it beneficial to have a use case + example of a sync scenario which is not possible by using the `offsetHelper`.
- "Values greater than 1 or less than 0 work fine." -> might be  good to explain this a little more/give an example with the two points + what they mean.

File: L.Map.Sync.js (hunk 2 above)
Comment (jjimenezshaw) [inline L.Map.Sync.js]: Sure!
.eslint was not complaining on the tests.

**Comment by jieter** [comment]
Interesting feature, thanks for the effort. I've added some comments and will have a look to the actual implementation later.
Might be interesting to record some GIFs and add them to the documentation.

File: L.Map.Sync.js (hunk 5 above)
Comment (jjimenezshaw) [inline L.Map.Sync.js]: While i was coding it I was thinking on where to place it. Yes, it is an "static" function (in Java or C++ notation), or a functions factory. It does not interact with the instance at all. That lets you more freedom to use your own function if you want something fancier.
However, it is going to be used only in this context, to create the function expected in offsetFn.

Maybe makes sense to move it to `L.Util.offsetHelper`... or with a different name?

**Comment by jjimenezshaw** [comment]
@jieter Thanks for your comments. I am open to new ways to explain the feature in the README file. Took me a long time writing it... but still I am not happy with the result. As I know how it works, I understand it. But I am sure it will be difficult for a new user. Feel free to suggest any change ;)

**Comment by jjimenezshaw** [comment]
@jieter When you say 

> Might be interesting to record some GIFs and add them to the documentation

do you mean an animated GIF on the README.md?

**Comment by jieter** [comment]
Yes, (small) animated GIFs in the readme.

**Comment by jjimenezshaw** [comment]
@jieter About different functions, I have just created one that syncs the maps in a global view. The maps will be displayed as a continuous one, doesn't matter where are the divs. If they overlap, they fit perfectly.
```JavaScript
function offsetGlobal(center, zoom, refMap, tgtMap) {                           
    var refC = refMap.getContainer();                                           
    var tgtC = tgtMap.getContainer();                                           
    var pt = refMap.project(center, zoom)                                       
                   .subtract([refC.offsetLeft, refC.offsetTop])                 
                   .subtract(refMap.getSize().divideBy(2))                       
                   .add([tgtC.offsetLeft, tgtC.offsetTop])                      
                   .add(tgtMap.getSize().divideBy(2));                          
    return refMap.unproject(pt, zoom);                                          
}        
```
Should I include it as part of the code (in `L.Util`) or just as an example in the README.md?

**Comment by jjimenezshaw** [comment]
@jieter I fixed all your comments (I hope), and added the animated gif (it is a bit small, I don't want to include a 2MB file just for that).

Apart from including or not `offsetGlobal` function in the code, I am done.

**Comment by jieter** [comment]
I'd say add it to the docs, not to the code. I'll merge this, you can open a new PR adding it to the docs.

**Comment by jieter** [comment]
0.1.0 is the latest of 6 releases
